First iteration on 2nd part blog post (#5)
damianavila/damian_blog: posts/draft2.md

# A deep dive into MyST, Part 2: TBD

> *This is a series of blog posts inviting you to join me in a little journey I have
experienced in the last few weeks at the time to figure it out a nice story for **MyST**

> inside the **Nikola** world.*

Reviewer: ```suggestion
> *This is a series of blog posts inviting you to join me in a little journey I have
experienced in the last few weeks to figure it out a nice story for **MyST**
inside the **Nikola** world.*
```

> In the previous blog [post](https://damianavila.github.io/blog/posts/a-deep-dive-into-myst-part-1-the-myst-parser-python-api-usage-in-nikola.html),

Reviewer: ```suggestion
In the previous [blog post](https://damianavila.github.io/blog/posts/a-deep-dive-into-myst-part-1-the-myst-parser-python-api-usage-in-nikola.html),
```

I think it's better to linkify at least two words at a time, otherwise it is easy for people to not realize it's a link

understand why roles and directives were not supported by the API as we expected.

In this post, we will explore the machinery underneath the MyST-Parser with the goal to
deepen our understanding and being able to propose some alternatives to provide the
expected support.

## Markdown-it-py and its plugins

Coming back to the MyST-Parser `default_parser` implementation, you can see the Parser
as a [collection of `markdown-it-py` plugins](https://mdit-py-plugins.readthedocs.io/en/latest/):

```python
    md = (
        MarkdownIt("commonmark", renderer_cls=renderer_cls)
        .enable("table")
        .use(front_matter_plugin)
        .use(myst_block_plugin)
        .use(myst_role_plugin)
        .use(footnote_plugin)
        .use(wordcount_plugin, per_minute=config.words_per_minute)
        .disable("footnote_inline")
        .disable("footnote_tail")
    )
```

The `MarkdownIt` class is instantiated with some parsing configuration options, dictating
the syntax rules and additional options for the parser and renderer. In addition,
several plugins are activated to load a collection of additional syntax rules and render
methods into the parser.

> When the input data is parsed via a nested chains of rules, it generates a list (stream)

Reviewer: ```suggestion
When the input data is parsed via nested chains of rules, it generates a list (stream)
```

        return publish_doctree(
            text,
            path.join(app.srcdir, docname + ".md"),
            reader=reader,
            parser=parser,
            parser_name="markdown",
            settings_overrides={"env": app.env, "gettext_compact": True},
        )
```

Finally, those objects are passed to the built-in Sphinx machinery to write the html output.

## The big question

This is great! We now understand why we need Sphinx to generate the expected HTML output
for roles and directives. Several questions now arise:

* Do we want to have Sphinx-free support for rendering roles and directives in the MyST-Parser 

Nikola (among other projects) already have their own machinery (based in Docutils) to
build the final HTML output. Getting the docutils object from the Python API would be
a super interesting way to easily expose and provide that object to the downstream
projects!

One caveat with this approach would be missing some Sphinx features (ie. cross-linking
between documents) based on custom roles and directives that we may need to re-implement.

2. Do we want to have docutils-free support for roles and directives in the MyST Python API?

> Docutils actually introduces the roles and directives concept (that Sphinx extend) so if

Reviewer: I think you need to indent the subsequent sentences in a markdown numbered list, or each number will start again at `1`

Author: Done!

we want to go docutils-free, then we will need to re-implement those concepts.

> 3. Does it makes sense to create a docutils alternative in Python? At what extend?

Reviewer: I'm not sure what you mean by "at what extend"

Author: The idea is to ask ourselves if we decide to go with a docutils alternative in Python, how much of the "docutils" functionality we may want to develop... Any suggestion about how to word that question in a short way?

> 3. Does it makes sense to create a docutils alternative in Python? At what extend?

Reviewer: ```suggestion
3. Does it makes sense to create a docutils alternative in Python? How much of its functionality would need to be replicated? How should it be extended or enhanced?
```

Reviewer: how about that?

(also, very intriguing question that I am a little bit terrified of, but is a very important question!)

Author: I like that proposal, and it captures the spirit of what I was asking for with the "At what extend?"

There is currently a nice example about an alternative implementation from the
Executable Books community, but in the Javascript world: https://github.com/executablebooks/markdown-it-docutils ;-).
Since we do not have Docutils there, it actually makes a lot of sense to write something
new. But, what is the value/need/place for an alternative implementation in Python?
Maybe, we do not need the whole Docutils stuff... but we maybe need some core
functionality?

If we decide to write some minimal support, what pieces are we interested to bring first?
Where those pieces should end up? The `markdown-it-docutils` package I referenced above is
actually a `markdown-it` (JS) plugin. If we follow that pattern, we should create a new

> `markdown-it-py-docutils` plugin and we are not longer in the MyST-Parser territory.

Reviewer: ```suggestion
`markdown-it-py-docutils` plugin and we are no longer in the MyST-Parser territory.
```

Author: Change ported in the last commit.
